### PR TITLE
cmd/dockerd: gracefully shut down the API server

### PR DESCRIPTION
**Pull request by corhere**
- Relates to #45057 

**- What I did**
Gave in-flight API requests up to five seconds to complete after all containers have been shut down before the daemon process exits.

**- How I did it**
By leveraging bleeding-edge APIs...from Go 1.8.

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: cmd/dockerd/daemon.go [hunk 1]
```diff
@@ -243,10 +256,30 @@ func (cli *DaemonCli) start(opts *daemonOptions) (err error) {
...
-		logrus.WithError(errAPI).Error("ServeAPI error")
+	var (
+		apiWG  sync.WaitGroup
+		errAPI = make(chan error, 1)
+	)
+	for _, ls := range lss {
+		apiWG.Add(1)
+		go func(ls net.Listener) {
+			defer apiWG.Done()
+			logrus.Infof("API listen on %s", ls.Addr())
```
Comment (thaJeztah) [inline cmd/dockerd/daemon.go:267]: I guess we're not using `WithField("listener", ls.Addr())` here to keep the message the same as before. (users _may_ be grepping for this). We could (separate from this PR) decide to change that, and use structured logging.

File: cmd/dockerd/daemon.go [hunk 2]
```diff
@@ -647,17 +675,20 @@ func checkTLSAuthOK(c *config.Config) bool {
 	return true
 }
 
-func loadListeners(cli *DaemonCli, tlsConfig *tls.Config) ([]string, error) {
-	if len(cli.Config.Hosts) == 0 {
-		return nil, errors.New("no hosts configured")
+func loadListeners(cfg *config.Config, tlsConfig *tls.Config) ([]net.Listener, []string, error) {
```
Comment (thaJeztah) [inline cmd/dockerd/daemon.go:678]: ❤️ we really need to make a pass at some of these. Too many places where we just pass everything (mostly out of convenience)

**Review by thaJeztah — COMMENTED** [review]
love it!! overall LGTM. left some comments / thoughts / suggestions; all relatively minor, but perhaps there's some you think make sense to include

File: cmd/dockerd/trap/trap.go [hunk 3]
```diff
@@ -4,47 +4,38 @@ import (
...
+// a new goroutine.
+//
+// If SIGINT or SIGTERM are received 3 times, the process is terminated
+// immediately (allows force quit of stuck daemon).
 func Trap(cleanup func(), logger interface {
 	Info(args ...interface{})
 }) {
-	c := make(chan os.Signal, 1)
-	// we will handle INT, TERM here
+	c := make(chan os.Signal, 10)
```
[the remark this answers is not in the document]
Comment (corhere) [reply cmd/dockerd/trap/trap.go:22]: The exact value is rather arbitrary, though I'll add a comment explaining why I chose a value > 1.

File: cmd/dockerd/trap/trap.go [hunk 4]
```diff
@@ -4,47 +4,38 @@ import (
...
 	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
```
[the remark this answers is not in the document]
Comment (corhere) [reply cmd/dockerd/trap/trap.go:23]: `signal.Notify(c, os.Interrupt)` also has special meaning on Windows. https://pkg.go.dev/os/signal?GOOS=windows#hdr-Windows

[the remark this answers is not in the document]
Comment (thaJeztah) [reply cmd/dockerd/trap/trap.go:23]: Yes, I was expecting to see windows having its own value, but it looks like it does _not_; Windows uses the POSIX variant; https://github.com/golang/go/blob/3f747d09e441999815fbb0a6025ade0f528967a8/src/os/exec_posix.go#L5
so on Windows it's also an alias for `syscall.SIGINT`

[the remark this answers is not in the document]
Comment (thaJeztah) [reply cmd/dockerd/trap/trap.go:23]: Oh, wait, it's not an alias...

File: cmd/dockerd/trap/trap.go [hunk 5]
```diff
@@ -4,54 +4,38 @@ import (
...
+// If SIGINT or SIGTERM are received 3 times, the process is terminated
+// immediately (allows force quit of stuck daemon).
 func Trap(cleanup func(), logger interface {
 	Info(args ...interface{})
 }) {
-	c := make(chan os.Signal, 1)
-	// we will handle INT, TERM, SIGPIPE here
-	signal.Notify(c, os.Interrupt, syscall.SIGTERM, syscall.SIGPIPE)
+	c := make(chan os.Signal, 10)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
```
Comment (cpuguy83) [inline cmd/dockerd/trap/trap.go:23]: Maybe `signal.NotifyContext` and we can pass around the context?

Comment (corhere) [reply cmd/dockerd/trap/trap.go:23]: `signal.Notify` would still be needed to implement the "force-quit on three ^C" behaviour.

**Review by vvoland — APPROVED** [review]

**Review by thaJeztah — APPROVED** [review]
LGTM

**Pull request merged**